x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -710,44 +756,53 @@ private void indexInvalidation(Collection<String> tokenIds, SecurityIndexManager
      */
-    public void refreshToken(String refreshToken, ActionListener<Tuple<UserToken, String>> listener) {
+    public void refreshToken(String refreshToken, ActionListener<Tuple<String, String>> listener) {
         ensureEnabled();
         final Instant refreshRequested = clock.instant();
         final Iterator<TimeValue> backoff = DEFAULT_BACKOFF.iterator();
         findTokenFromRefreshToken(refreshToken,
             backoff,
             ActionListener.wrap(tokenDocHit -> {
                 final Authentication clientAuth = Authentication.readFromContext(client.threadPool().getThreadContext());
-                innerRefresh(tokenDocHit.getId(), tokenDocHit.getSourceAsMap(), tokenDocHit.getSeqNo(), tokenDocHit.getPrimaryTerm(),
-                        clientAuth, backoff, refreshRequested, listener);
+                innerRefresh(refreshToken, tokenDocHit.getId(), tokenDocHit.getSourceAsMap(), tokenDocHit.getSeqNo(),
+                    tokenDocHit.getPrimaryTerm(),
+                    clientAuth, backoff, refreshRequested, listener);
             }, listener::onFailure));
     }
 
     /**
-     * Inferes the format and version of the passed in {@code refreshToken}. Delegates the actual search of the token document to
+     * Infers the format and version of the passed in {@code refreshToken}. Delegates the actual search of the token document to
      * {@code #findTokenFromRefreshToken(String, SecurityIndexManager, Iterator, ActionListener)} .
      */
     private void findTokenFromRefreshToken(String refreshToken, Iterator<TimeValue> backoff, ActionListener<SearchHit> listener) {
-        if (refreshToken.length() == TOKEN_ID_LENGTH) {
+        if (refreshToken.length() == TOKEN_LENGTH) {
             // first check if token has the old format before the new version-prepended one
             logger.debug("Assuming an unversioned refresh token [{}], generated for node versions"
-                    + " prior to the introduction of the version-header format.", refreshToken);
+                + " prior to the introduction of the version-header format.", refreshToken);
             findTokenFromRefreshToken(refreshToken, securityMainIndex, backoff, listener);
         } else {
-            try {
-                final Tuple<Version, String> versionAndRefreshTokenTuple = unpackVersionAndPayload(refreshToken);
-                final Version refreshTokenVersion = versionAndRefreshTokenTuple.v1();
-                final String unencodedRefreshToken = versionAndRefreshTokenTuple.v2();
-                if (false == refreshTokenVersion.onOrAfter(VERSION_TOKENS_INDEX_INTRODUCED)
-                        || unencodedRefreshToken.length() != TOKEN_ID_LENGTH) {
-                    logger.debug("Decoded refresh token [{}] with version [{}] is invalid.", unencodedRefreshToken, refreshTokenVersion);
+            if (refreshToken.length() == HASHED_TOKEN_LENGTH) {
```

Comment:
When is this if branch considered?

Reply:
This is considered during IDP initiated SAML logout when we need to find relevant tokens based on the token metadata ( that contain the SAML NameID ) and then invalidate those explicitly. See https://github.com/elastic/elasticsearch/blob/9beb31fd3c5a8323cb08cc524f1a2268e9c72c24/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlInvalidateSessionAction.java